File: Experiment1/src/scripts_with-g＿20250204/ghidra/automation_20250206/scripts/comparePcode.py
import os
import re
import csv
from difflib import SequenceMatcher

def normalize_line(line):
    # 將所有 0x… 數字替換成 0xX，避免因數值不同而影響比較
    line = re.sub(r'0x[0-9a-fA-F]+', '0xX', line)
    # 可以加上其他規則，例如移除多餘空白
    return line.strip()

def get_pcode_content(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    marker = "High-level P-code:"
    idx = content.find(marker)
    if idx != -1:
        # 取 marker 之後的內容
        pcode = content[idx + len(marker):].strip()
        # 將每一行正規化
        lines = [normalize_line(line) for line in pcode.splitlines() if line.strip()]
        return "\n".join(lines)
    return ""

def similarity(s1, s2):
    return SequenceMatcher(None, s1, s2).ratio()

def read_map(dir_path):
    functions = {}
    map_path = os.path.join(dir_path, "function_map.csv")
    with open(map_path, newline='', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            label = row["Label"]
            function = row["Function"]
            source_file = row["SourceFile"]
            arch = row["Architecture"]
            # 檔案命名規則: label_<Label>_<Function>_pcode.txt
            pcode_filename = f"label_{label}_{function}_pcode.txt"
            pcode_path = os.path.join(dir_path, pcode_filename)
            if os.path.exists(pcode_path):
                pcode_content = get_pcode_content(pcode_path)
            else:
                pcode_content = ""
            functions[label] = {
                "Function": function,
                "SourceFile": source_file,
                "Architecture": arch,
                "Pcode": pcode_content
            }
    return functions

def count_pcode_lines(pcodeContent):
    if pcodeContent:
        return len(pcodeContent.splitlines())
    return 0

def calculate_match_ratio(arm_content, mips_content):
    return similarity(arm_content, mips_content) * 100

def generate_comparison_notes(armLines, mipsLines, matchRatio):
    notes = []
    if armLines > 0 and abs(armLines - mipsLines) > armLines * 0.2:
        notes.append("P-code行數差異顯著")
    if matchRatio < 50:
        notes.append("低相似度")
    elif matchRatio > 90:
        notes.append("高相似度")
    return "; ".join(notes)

def main():
    # 修改這裡的路徑以符合您的目錄
    arm_dir = "/home/tommy/cross-architecture/Experiment1/src/scripts_with-g＿20250204/ghidra/automation_20250206/results/analysis_20250206_165530/arm"
    mips_dir = "/home/tommy/cross-architecture/Experiment1/src/scripts_with-g＿20250204/ghidra/automation_20250206/results/analysis_20250206_165530/mips"
    
    arm_functions = read_map(arm_dir)
    mips_functions = read_map(mips_dir)
    
    output_csv = "comparison_results.csv"
    with open(output_csv, "w", newline='', encoding='utf-8') as csvfile:
        fieldnames = ["Label", "Function", "SourceFile", "ARM_PcodeLines", "MIPS_PcodeLines", "Match_Ratio", "Notes"]
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()
        
        all_labels = set(arm_functions.keys()).union(mips_functions.keys())
        for label in all_labels:
            arm_info = arm_functions.get(label)
            mips_info = mips_functions.get(label)
            if arm_info and mips_info:
                arm_lines = count_pcode_lines(arm_info["Pcode"])
                mips_lines = count_pcode_lines(mips_info["Pcode"])
                match_ratio = calculate_match_ratio(arm_info["Pcode"], mips_info["Pcode"])
                notes = generate_comparison_notes(arm_lines, mips_lines, match_ratio)
                
                writer.writerow({
                    "Label": label,
                    "Function": arm_info["Function"],
                    "SourceFile": arm_info["SourceFile"],
                    "ARM_PcodeLines": arm_lines,
                    "MIPS_PcodeLines": mips_lines,
                    "Match_Ratio": f"{match_ratio:.2f}",
                    "Notes": notes
                })
    print("比對完成，結果保存至:", output_csv)

if __name__ == "__main__":
    main()

File: Experiment1/src/scripts_with-g＿20250204/ghidra/automation_20250206/scripts/ComparePcode.java
import java.io.*;
import java.util.*;

public class ComparePcode {
    static class FunctionInfo {
        String name;
        String sourceFile;
        String architecture;
        String pcodeContent;
        
        FunctionInfo(String name, String sourceFile, String architecture, String pcodeContent) {
            this.name = name;
            this.sourceFile = sourceFile;
            this.architecture = architecture;
            this.pcodeContent = pcodeContent;
        }
    }
    
    public static void main(String[] args) {
        if (args.length != 3) {
            System.out.println("Usage: java ComparePcode <arm_dir> <mips_dir> <output_csv>");
            return;
        }
        
        String armDir = args[0];
        String mipsDir = args[1];
        String outputCsv = args[2];
        
        try {
            // 讀取函數映射
            Map<Integer, FunctionInfo> armFunctions = readFunctionMap(armDir);
            Map<Integer, FunctionInfo> mipsFunctions = readFunctionMap(mipsDir);
            
            // 創建比對結果文件
            PrintWriter writer = new PrintWriter(new FileWriter(outputCsv));
            writer.println("Label,Function,SourceFile,ARM_PcodeLines,MIPS_PcodeLines,Match_Ratio,Notes");
            
            // 取得所有標籤
            Set<Integer> allLabels = new HashSet<>();
            allLabels.addAll(armFunctions.keySet());
            allLabels.addAll(mipsFunctions.keySet());
            
            for (Integer label : allLabels) {
                FunctionInfo armInfo = armFunctions.get(label);
                FunctionInfo mipsInfo = mipsFunctions.get(label);
                
                if (armInfo != null && mipsInfo != null) {
                    int armLines = countPcodeLines(armInfo.pcodeContent);
                    int mipsLines = countPcodeLines(mipsInfo.pcodeContent);
                    double matchRatio = calculateMatchRatio(armInfo.pcodeContent, mipsInfo.pcodeContent);
                    
                    String notes = generateComparisonNotes(armLines, mipsLines, matchRatio);
                    
                    writer.printf("%d,%s,%s,%d,%d,%.2f,%s%n",
                                label, armInfo.name, armInfo.sourceFile,
                                armLines, mipsLines, matchRatio, notes);
                }
            }
            
            writer.close();
            System.out.println("比對完成，結果保存至: " + outputCsv);
            
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    
    private static Map<Integer, FunctionInfo> readFunctionMap(String dir) throws IOException {
        Map<Integer, FunctionInfo> functions = new HashMap<>();
        
        File mapFile = new File(dir, "function_map.csv");
        BufferedReader reader = new BufferedReader(new FileReader(mapFile));
        // 跳過標題行
        reader.readLine();
        
        String line;
        while ((line = reader.readLine()) != null) {
            String[] parts = line.split(",");
            int label = Integer.parseInt(parts[0]);
            String functionName = parts[1];
            String sourceFile = parts[2];
            String architecture = parts[3];
            
            // 讀取對應的 P-code 文件
            String pcodeFile = String.format("%s/label_%d_%s_pcode.txt", dir, label, functionName);
            String pcodeContent = readPcodeFile(pcodeFile);
            
            functions.put(label, new FunctionInfo(functionName, sourceFile, architecture, pcodeContent));
        }
        
        reader.close();
        return functions;
    }
    
    private static String readPcodeFile(String filename) throws IOException {
        StringBuilder content = new StringBuilder();
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        
        String line;
        boolean isPcode = false;
        while ((line = reader.readLine()) != null) {
            if (line.equals("P-code:")) {
                isPcode = true;
                continue;
            }
            if (isPcode && !line.trim().isEmpty()) {
                content.append(line).append("\n");
            }
        }
        
        reader.close();
        return content.toString();
    }
    
    private static int countPcodeLines(String pcodeContent) {
        return pcodeContent.split("\n").length;
    }
    
    // 調整後的匹配率計算：對每一行計算一個 0~1 的相似度，並取平均
    private static double calculateMatchRatio(String arm, String mips) {
        String[] armLines = arm.split("\n");
        String[] mipsLines = mips.split("\n");
        
        int total = Math.max(armLines.length, mipsLines.length);
        int min = Math.min(armLines.length, mipsLines.length);
        double sumSimilarity = 0.0;
        
        for (int i = 0; i < min; i++) {
            double sim = computeSimilarity(armLines[i].trim(), mipsLines[i].trim());
            sumSimilarity += sim;
        }
        // 假設超出的行視為 0 相似度
        return (sumSimilarity / total) * 100;
    }
    
    // 計算兩個字串的相似度，基於 Levenshtein 距離的歸一化結果
    private static double computeSimilarity(String s1, String s2) {
        int distance = levenshteinDistance(s1, s2);
        int maxLen = Math.max(s1.length(), s2.length());
        if (maxLen == 0) return 1.0;
        return 1.0 - ((double) distance / maxLen);
    }
    
    // 標準的 Levenshtein 距離實作
    private static int levenshteinDistance(String s1, String s2) {
        int len1 = s1.length();
        int len2 = s2.length();
        int[][] dp = new int[len1 + 1][len2 + 1];
        
        for (int i = 0; i <= len1; i++) {
            dp[i][0] = i;
        }
        for (int j = 0; j <= len2; j++) {
            dp[0][j] = j;
        }
        
        for (int i = 1; i <= len1; i++) {
            for (int j = 1; j <= len2; j++) {
                int cost = s1.charAt(i - 1) == s2.charAt(j - 1) ? 0 : 1;
                dp[i][j] = Math.min(
                        Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1),
                        dp[i - 1][j - 1] + cost);
            }
        }
        return dp[len1][len2];
    }
    
    private static String generateComparisonNotes(int armLines, int mipsLines, double matchRatio) {
        StringBuilder notes = new StringBuilder();
        
        if (Math.abs(armLines - mipsLines) > armLines * 0.2) {
            notes.append("P-code行數差異顯著; ");
        }
        
        if (matchRatio < 50) {
            notes.append("低相似度; ");
        } else if (matchRatio > 90) {
            notes.append("高相似度; ");
        }
        
        return notes.toString().trim();
    }
}
